### PR TITLE
feat: fetch LLM API keys from user env variables

### LINKED ISSUE #76: fix(llm): address an issue where saved AI API keys are not carried over to the next sessions
## Reproducible Steps

1. Create any plots and display via `maidr.show()`

1. Hit H key and provide API keys for both OpenAI and Google Gemini

1. Click Save and Close

1. If prompted, save the password in your browser

1. Open LLM from the interactive plot area via Ctrl+Shift+/ (on Windows) or Alt+Shift+/ (on Mac)

1. Make sure the AI responses are working

1. Close the maidr browser and exit out of the current Python repl

1. Repeat the steps above and see if the AI API keys are preserved to the next sessions

## Current Behavior

API keys are not carried over to the next sessions even when it's saved in the browser

## Suggested Solution

1. When `maidr.show` is executed search for the following env variables via `os.getenv()`:

* `OPENAI_API_KEY`

* `GOOGLE_API_KEY`

1. Use these keys as fallback

### PR DESCRIPTION
**Pull request by dakshpokar**
## Description
This pull request fixes the handling of API keys for LLMs in the code. It adds a JavaScript script to handle the API keys for LLMs and initializes the LLM secrets in the MAIDR instance. The script injects the LLM API keys into the MAIDR instance and sets the appropriate settings based on the presence of the Gemini and OpenAI API keys. This ensures that the LLM functionality works correctly with the updated API key handling.

closes #76 

## Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

# Pull Request

## Description
1. Added a new method called `initialize_llm_secrets()` in environment.py which fetches the keys from the environment variable.
2. Injected the script when the maidr iframe loads initially.

## Checklist

- [x] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [x] I have performed a self-review of my own code and ensured it follows the project's coding standards.
- [x] I have tested the changes locally following `ManualTestingProcess.md`, and all tests related to this pull request pass.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation, if applicable.
- [x] I have added appropriate unit tests, if applicable.

### CODE REVIEW

**Comment by dakshpokar** [comment]
Greetings Professor @jooyoungseo,

While this PR addresses the issue of LLM keys being lost due to changes in the port number each time a Maidr instance is run, I believe we should consider a more scalable solution, especially in light of potential future changes. I discussed this matter with @SaaiVenkat, and he shares the concern that these fixes do not address the core problem but rather serve as a workaround for a dependency issue with the library. This issue could have been more easily resolved if `py-htmltools` provided a way to fix the port.

In my opinion, a more sustainable solution would be to implement a custom server to serve our HTML files, rather than relying on py-htmltools for this functionality. However, this is just my suggestion, and I would appreciate your thoughts on the matter.

Best Regards,
Daksh Pokar

**Comment by jooyoungseo** [comment]
I had a Slack communication with @dakshpokar  and we agreed that we need to use a custom local server and pin down a port number for an interactive console sessions (e.g., vanilla Python repl; ipython repl) while keeping the current behavior of `htmltools.show()` for shiny and any other web widgets. This way, we can preserve the same API keys across different sessions. Of course, this may be a stopgap solution and there might be a better way to handle this case, but we want to patch this first for the sake of the importance of this issue against our end-users.

File: maidr/core/maidr.py [hunk 1]
```diff
@@ -1,18 +1,17 @@
 from __future__ import annotations
 
-from typing import Literal
```
Comment (SaaiVenkat) [inline maidr/core/maidr.py:3]: Why is this changed?

File: maidr/core/maidr.py [hunk 2]
```diff
@@ -1,18 +1,17 @@
...
-
 import io
 import json
 import uuid
+from typing import Literal
 
 from htmltools import HTML, HTMLDocument, Tag, tags
 from lxml import etree
-
```
Comment (SaaiVenkat) [inline maidr/core/maidr.py:11]: Why is this line removed?

File: maidr/core/maidr.py [hunk 3]
```diff
@@ -1,18 +1,17 @@
 from __future__ import annotations
 
-from typing import Literal
-
 import io
 import json
 import uuid
+from typing import Literal
```
Comment (SaaiVenkat) [inline maidr/core/maidr.py:6]: Please revert this

**Comment by SaaiVenkat** [comment]
> I had a Slack communication with @dakshpokar and we agreed that we need to use a custom local server and pin down a port number for an interactive console sessions (e.g., vanilla Python repl; ipython repl) while keeping the current behavior of `htmltools.show()` for shiny and any other web widgets. This way, we can preserve the same API keys across different sessions. Of course, this may be a stopgap solution and there might be a better way to handle this case, but we want to patch this first for the sake of the importance of this issue against our end-users.

Professor @jooyoungseo , since this is stopgap solution, I would suggest to create a backlog ticket right away to address this issue and link it with this issue.

cc: @dakshpokar

**Comment by jooyoungseo** [comment]
Update: @dakshpokar and I decided to make a separate PR addressing the custom local server implementation. However, I vote for the current approach where users save their API keys in their env variables (e.g., `OPENAI_API_KEY`; `GEMINI_API_KEY`) for the recurrent and stable use across different sessions. This approach also aligns with other major libs (e.g., openai; langchain). I will merge this PR today and will create another issue for the custom server.

**Comment by jooyoungseo** [comment]
@dakshpokar `GEMINI_API_KEY` seems to be an officially recommended key name: https://github.com/google-gemini/generative-ai-python

File: maidr/util/environment.py [hunk 4]
```diff
@@ -11,3 +15,66 @@ def is_interactive_shell() -> bool:
...
+                        iframe.contentWindow.localStorage.setItem(key, value);
+                    }} catch (error) {{
+                        console.error('Error accessing iframe localStorage:', error);
+                    }}
+                }} else {{
+                    console.error('Iframe not found or inaccessible.');
+                }}
+            }}
+            addKeyValueLocalStorage(
+                'maidr-iframe', 'settings_data', JSON.stringify({settings_data})
```
Comment (jooyoungseo) [inline maidr/util/environment.py:76]: Do NOT use the same `iframeId` here. Based on my testing result, the user env variable change does not get picked up dynamically once a setting is stored to this iframe. since you have pinned down the iframeId as "maidr-iframe-id,"the same LLM setting is preserved regardless of user's env variable update. Use a random id just like https://github.com/dakshpokar/py_maidr/blob/2e08b9d7803c7d544e897cc4b063a5af0f473b8b/maidr/core/maidr.py#L145-L149

However, I still suggest you bind this local store handler to maidr object instead of iframe.

File: maidr/util/environment.py [hunk 5]
```diff
@@ -11,3 +15,66 @@ def is_interactive_shell() -> bool:
...
+            settings["openAIAuthKey"] = openai_api_key
+            settings["LLMOpenAiMulti"] = True
+            settings["LLMGeminiMulti"] = True
+            settings["LLMModel"] = "multi"
+        elif openai_api_key is not None:
+            settings["LLMOpenAiMulti"] = True
+            settings["openAIAuthKey"] = openai_api_key
+        elif gemini_api_key is not None:
+            settings["LLMGeminiMulti"] = True
+            settings["geminiAuthKey"] = gemini_api_key
```
Comment (jooyoungseo) [inline maidr/util/environment.py:58]: Only when gemini key is available, you need to set to gemini. If we don't this, this will fall back to the default "openai" model.

```suggestion
        elif gemini_api_key is not None:
            settings["LLMGeminiMulti"] = True
            settings["geminiAuthKey"] = gemini_api_key
            settings["LLMModel"] = "gemini"
```

File: maidr/util/environment.py (hunk 4 above)
Comment (jooyoungseo) [reply maidr/util/environment.py:76]: I have tested this again and the env variable update gets picked up as expected even with this fixed id. However, please use `class` instead of `id` if you need to go with this tweak. For `id`, assign a random unique iframe id if needed.

File: maidr/util/environment.py [hunk 6]
```diff
@@ -11,3 +15,66 @@ def is_interactive_shell() -> bool:
...
+
+        if gemini_api_key is not None and openai_api_key is not None:
+            settings["geminiAuthKey"] = gemini_api_key
+            settings["openAIAuthKey"] = openai_api_key
+            settings["LLMOpenAiMulti"] = True
+            settings["LLMGeminiMulti"] = True
+            settings["LLMModel"] = "multi"
+        elif openai_api_key is not None:
+            settings["LLMOpenAiMulti"] = True
+            settings["openAIAuthKey"] = openai_api_key
```
Comment (jooyoungseo) [inline maidr/util/environment.py:55]: Specifying "openai" is safer as the default value might change later.

```suggestion
        elif openai_api_key is not None:
            settings["LLMOpenAiMulti"] = True
            settings["openAIAuthKey"] = openai_api_key
            settings["LLMModel"] = "openai"
```

File: maidr/util/environment.py (hunk 4 above)
Comment (dakshpokar) [reply maidr/util/environment.py:76]: Sure Professor @jooyoungseo, making this change!

File: maidr/util/environment.py (hunk 5 above)
Comment (dakshpokar) [reply maidr/util/environment.py:58]: Thank you Professor @jooyoungseo, will check this!

File: maidr/util/environment.py (hunk 6 above)
Comment (dakshpokar) [reply maidr/util/environment.py:55]: Sure Professor @jooyoungseo, making this change!

File: maidr/util/environment.py (hunk 4 above)
Comment (dakshpokar) [reply maidr/util/environment.py]: Professor, @jooyoungseo. I have assigned a unique id to the maidr iframe. I could've used class but it returns an array of matched elements which I have to then loop over to add the key values.

File: maidr/core/maidr.py [hunk 7]
```diff
@@ -170,8 +169,11 @@ def _inject_plot(plot: HTML, maidr: str) -> Tag:
             tags.script(maidr),
         )
 
+        unique_id = Maidr._unique_id()
+
```
Comment (jooyoungseo) [inline maidr/core/maidr.py:173]: Are you recycling the same unique_id used for maidr object towards iframe?This may cause some issues in the long run because there are two matches per id (one for iframe; the other for maidr). What do you think about adding a prefix (e.g., "iframe_") to distinguish the iframe id from the maidr one?

File: maidr/util/environment.py [hunk 8]
```diff
@@ -11,3 +15,68 @@ def is_interactive_shell() -> bool:
             )
         except ImportError:
             return False
+
+    @staticmethod
+    def initialize_llm_secrets(unique_id: str) -> str:
```
Comment (jooyoungseo) [inline maidr/util/environment.py:20]: Maybe I could be wrong, but don't we use `_` prefix for static method?

Comment (dakshpokar) [reply maidr/util/environment.py:20]: Professor @jooyoungseo, it is generally used for the methods that are private or internal.

File: maidr/core/maidr.py (hunk 7 above)
Comment (dakshpokar) [reply maidr/core/maidr.py:173]: Professor @jooyoungseo, as the method generates a unique uuid everytime the `Maidr._unique_id()` is called, the conflict might not arise, however, to be on a safer side I will add a prefix iframe_ for the id.

**Review by jooyoungseo — APPROVED** [review]
I have tested all the possible edge cases with various env variable combinations. It is working very reliably. LGTM! I am merging it now.
